### PR TITLE
onError and retry definition updates

### PR DESCRIPTION
**Pull request by tsurdilo**
Clearly defines onError and retry definitions for states.
retry only allowed in Event, Operation, Parallel, and ForEach states. 
Added examples on how to use onError, retry, and combination of both for powerful error handling within states.

### CODE REVIEW

**Comment by tsurdilo** [comment]
Related to Issue https://github.com/cncf/wg-serverless/issues/164

**Comment by tsurdilo** [comment]
@manuelstein would you mind taking a quick look please? Thanks.

**Comment by manuelstein** [comment]
Alright, thank you for preparing the PR

1. Workflow runtime error handling ($.onError) is removed as global configuration.
This was not defining default behaviour but handling of a **workflow runtime error**. As this is a global error (the workflow runtime has entirely failed, behaviour undefined), there is no point to make a transition anymore.

2. Retry has been taken out of the action definition and instead can be specified for **event** state, **operation** state, **parallel** state and **foreach** state, because only these state types carry actions. The execution of a remote insecure/third-party action may fail due to temporary errors that are "retry-able", e.g. unavailability of some kind.
The **subflow** state doesn't get a retry behaviour because it may run (and fail) independently with the calling state not waiting on its completion (or failure)

3. correction in errordatafilter description

Only one suggestion: The PR changes the retry definition's *expression* to *errorExpression*, maybe that should remain *expression*.

File: workflow/spec/schema/serverless-workflow-schema-01.json [hunk 1]
```diff
@@ -259,27 +243,23 @@
       "type": "object",
       "description": "Retry Definition",
       "properties": {
-        "expression": {
-          "description": "Boolean expression that matches against the function results. Must be evaluated to true for retry policy to trigger",
+        "errorExpression": {
+          "description": "Expression that matches against error properties",
```
Comment (manuelstein) [inline workflow/spec/schema/serverless-workflow-schema-01.json]: The retry definition specifies an expression to match the state's output against to check whether it needs to be retried. I could imagine that also non-error outcomes may cause retries, e.g. to monitor an external operation. A complex action that retrieves a status may return with "PENDING" and the workflow may want to wait until the returned status is "COMPLETE", which doesn't involve any errors.

**Comment by tsurdilo** [comment]
@manuelstein thanks for comments - updated to rename errorExpression to expression as you mentioned. Also defined that this expression is evaluated against the states data output so that your use case in second comment can be accomplished as well.

**Comment by manuelstein** [comment]
LGTM

**Comment by tsurdilo** [comment]
rebased

**Pull request merged**